rir/src/compiler/opt/hoist_instruction.cpp

```diff
@@ -0,0 +1,187 @@
+                           dom.dominates(target->next0, bb) &&
+                           (!target->next1 || dom.dominates(target->next1, bb)))
+                        target = target->next0;
+            }
+
+            if (!target) {
+                ip = next;
+                continue;
+            }
+
+            auto doesNotReorderEffects = [&](BB* x) {
+                std::function<bool(BB*)> compute = [&](BB* x) {
+                    if (!x || x == target)
+                        return true;
+                    // done, check cur bb for effects
+                    if (x == bb) {
+                        for (auto& j : *bb) {
+                            if (i == j)
+                                return true;
+                            if (j->hasStrongEffects())
+                                return false;
+                        }
+                        assert(false);
+                    }
+                    if (!x->isEmpty()) {
+                        // We can only hoist effects over branches if both
+                        // branch targets will trigger the effect
+                        if (x->last()->branches()) {
+                            if (!dom.dominates(x->trueBranch(), bb) ||
+                                !dom.dominates(x->falseBranch(), bb))
+                                return false;
+                        }
+                    }
+
+                    for (auto& j : *bb)
+                        if (j->hasStrongEffects())
+                            return false;
+
+                    return compute(x->next0) && compute(x->next1);
+                };
+                return compute(x->next0) && compute(x->next1);
+            };
+
+            auto noUnneccessaryComputation = [&](BB* x, unsigned exceptions) {
+                std::function<bool(BB*)> compute = [&](BB* x) {
+                    if (!x || x == bb || x == target)
+                        return true;
+                    if (!x->isEmpty()) {
+                        // If we hoist over a branch, but one of the
+                        // branches does not need the value, then this will
+                        // waste computation
+                        if (x->last()->branches()) {
+                            if (!dom.dominates(x->trueBranch(), bb) ||
+                                !dom.dominates(x->falseBranch(), bb)) {
+                                if (exceptions == 0)
+                                    return false;
+                                exceptions--;
+                            }
+                        }
+                    }
+                    return compute(x->next0) && compute(x->next1);
+                };
+                return compute(x->next0) && compute(x->next1);
+            };
+
+            bool success = true;
+            if (i->hasObservableEffects())
+                success = doesNotReorderEffects(target);
+            else if (i->cost() > 0)
```

Comment:
Why we have a blacklist but then a cost for ruling out `castType`?

Reply:
it's the other way round. castType is free (since it's not an actual instruction) so we can always hoist it (success is always true)